pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/util/autocomplete/AstPackageExplorer.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.util.fxdesigner.util.autocomplete;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Modifier;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import net.sourceforge.pmd.lang.Language;
+
+
+/**
+ * Finds XPath node names by looking into the classpath
+ * directory corresponding to the AST of a language. This
+ * is ok for Java, Apex, etc. but not e.g. for XML.
+ */
+class AstPackageExplorer implements NodeNameFinder {
+    private List<String> availableNodeNames = new ArrayList<>();
+
+
+    AstPackageExplorer(Language language) {
+        availableNodeNames =
+            getClasses("net.sourceforge.pmd.lang."
+                           + language.getTerseName() + ".ast")
+                .filter(clazz -> clazz.getSimpleName().startsWith("AST"))
+                .filter(clazz -> !clazz.isInterface() && !Modifier.isAbstract(clazz.getModifiers()))
+                .map(m -> m.getSimpleName().substring("AST".length()))
+                .collect(Collectors.toList());
+
+    }
+
+
+    @Override
+    public List<String> getNodeNames() {
+        return availableNodeNames;
+    }
+
+
+    /** Finds the classes in the given package by looking in the classpath directories. */
+    private static Stream<Class<?>> getClasses(String packageName) {
+
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        assert classLoader != null;
+
+        String path = packageName.replace('.', '/');
+
+        Enumeration<URL> resources;
+
+        try {
+            resources = classLoader.getResources(path);
+        } catch (IOException e) {
+            return Stream.empty();
+        }
+
+        List<File> dirs = new ArrayList<>();
+
+        while (resources.hasMoreElements()) {
+            URL resource = resources.nextElement();
+            dirs.add(new File(resource.getFile()));
+        }
+
+        List<Class<?>> classes = new ArrayList<>();
+
+        for (File directory : dirs) {
```

Comment:
doesn't `Files.walkTree` work here? or does that have a problem with this being inside a jar?

Reply:
I didn't know about that, but looking into it I'm not sure what it would achieve. We don't need to recurse in the subdirectories, there's no link to follow --it seems to me like overkill. Wdyt?

Reply:
Cool I'll try that!

Reply:
It actually doesn't change code size, but this may be easily extended later to e.g. get classes in a package and its subpackages so I like it. I'd like to see this moved into a core Util class someday